### PR TITLE
[ios] Add manual constraints workbook

### PR DESCRIPTION
**Pull request by VincentDondain**
Took a bit of https://developer.xamarin.com/guides/ios/user_interface/programmatic-layout-constraints / https://developer.apple.com/library/content/documentation/UserExperience/Conceptual/AutolayoutPG/ProgrammaticallyCreatingConstraints.html

The goal is to show key points when using constraints created via code like `TranslatesAutoresizingMaskIntoConstraints` and its impact.

I also believe that constraints are a great thing to use workbooks for because it's a complex but very visual concept.

It should encourage people to use workbooks when trying to setup their UI (:

Feedback welcomed.

### CODE REVIEW

File: ios/platform-features/auto-layout/programmatical-constraints.workbook [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+};
+
+// This could be any parent view you add UI elements to.
+var rootView = RootViewController.View;
+rootView.AddSubview (orangeLabel);
+```
+
+#### Absolute coordinates
+
+Let’s position the `orangeLabel` at the top left corner manualy first so we can see it.
```
Comment (asb3993) [inline ios/platform-features/auto-layout/programmatical-constraints.workbook]: manualy -> manually

File: ios/platform-features/auto-layout/programmatical-constraints.workbook [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+If we wanted to center this label in the middle of the screen we could, *technically*, do it using **absolute** x,y coordinates but this is **not** the most optimal, responsive and simple (will imply quite a bit of math) way to do it.
+
+#### Begin with Constaints
+
+`TranslatesAutoresizingMaskIntoConstraints` is **key** to using manual constraints.
+
+By default this is set to `true` which lets the system create a set of constraints for you.
+
+However it prevents you from adding **additional constraints**, see API doc:
```
Comment (asb3993) [inline ios/platform-features/auto-layout/programmatical-constraints.workbook]: Add a link to the API doc that you refer to here.

Comment (asb3993) [inline ios/platform-features/auto-layout/programmatical-constraints.workbook]: I'd probably alter the sentence a little to say

"However, using `TranslatesAutoresizingMaskIntoConstraints` prevents you from adding **additional constraints**, as described in [Specific API doc name] doc:"

File: ios/platform-features/auto-layout/programmatical-constraints.workbook (hunk 1 above)
Comment (asb3993) [inline ios/platform-features/auto-layout/programmatical-constraints.workbook]: manualy -> manually

File: ios/platform-features/auto-layout/programmatical-constraints.workbook (hunk 2 above)
Comment (asb3993) [inline ios/platform-features/auto-layout/programmatical-constraints.workbook]: Add a link to the API doc.

As it's not 100% clear what 'it' refers to, I'd rephrase this to say:

"However, using `TranslatesAutoresizingMaskIntoConstraints` prevents you from adding **additional constraints**, as described in [name of specific API doc] doc:

File: ios/platform-features/auto-layout/programmatical-constraints.workbook [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+> Note that the autoresizing mask constraints fully specify the view’s size and position; therefore, you cannot add additional constraints to modify this size or position without introducing conflicts. If you want to use Auto Layout to dynamically calculate the size and position of your view, you must set this property to false, and then provide a non ambiguous, nonconflicting set of constraints for the view.
+
+```csharp
+orangeLabel.TranslatesAutoresizingMaskIntoConstraints = false;
+```
+
+Setting `TranslatesAutoresizingMaskIntoConstraints` to `false` should result in the label disappearing because we did not, *yet*, add our manual constraints.
```
Comment (asb3993) [inline ios/platform-features/auto-layout/programmatical-constraints.workbook]: It didn't for me--although I think this is an issue with controls in a workbook. Maybe worth saying that 'if you were coding this in an app....'

File: ios/platform-features/auto-layout/programmatical-constraints.workbook [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+Finally we can restore the label’s **width** and **height**.
+
+```csharp
+orangeLabel.WidthAnchor.ConstraintEqualTo (120).Active = true;
+orangeLabel.HeightAnchor.ConstraintEqualTo (50).Active = true;
+```
+
+Feel free to play around with the constraints above to try to center the label vertically **and** horizontally.
+
+#### Purple label constrained to orange label
```
Comment (asb3993) [inline ios/platform-features/auto-layout/programmatical-constraints.workbook]: This isn't a big issue, but H4 renders super similar to text in workbooks. Maybe worth bumping up all your headings by one?

File: ios/platform-features/auto-layout/programmatical-constraints.workbook [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+Now you could, for instance, constrain the trailing anchor of the purple label to the trailing anchor of the marin guide to fill all the horizontal space.
+
+*Note that this will adapat to **any device form factor**.*
+
+```csharp
+purpleLabel.TrailingAnchor.ConstraintEqualTo (marginGuide.TrailingAnchor).Active = true;
+```
+
+A good exercise now would be to try positioning the labels vertically instead of horizontally.
```
Comment (asb3993) [inline ios/platform-features/auto-layout/programmatical-constraints.workbook]: Optional: Add a "Try it now" heading above this exercise

File: ios/platform-features/auto-layout/programmatical-constraints.workbook (hunk 2 above)
Comment (VincentDondain) [inline ios/platform-features/auto-layout/programmatical-constraints.workbook]: @asb3993 the documentation I got for `TranslatesAutoresizingMaskIntoConstraints` is from Apple: https://developer.apple.com/reference/uikit/uiview/1622572-translatesautoresizingmaskintoco

Ours is less clear on that: https://developer.xamarin.com/api/property/MonoTouch.UIKit.UIView.TranslatesAutoresizingMaskIntoConstraints/

We're just saying: "set this property to false provides a clear starting point for programmatically specifying constraints".

They are explaining **what** the autoresizing mask does and **why** you can't add additional constraints just like that _(would introduce conflicts)_ and then they say that: "if you **want** to use Auto Layout to **dynamically calculate** the size and position of your view, you **must** set this property to **false**".

It's much clearer, should we update our API doc so that I can reference it? Otherwise I'd rather like to reference Apple's :P

Comment (conceptdev) [inline ios/platform-features/auto-layout/programmatical-constraints.workbook]: let's use Apple's link for now, but also create an issue for API writers to update our doc (and link back here, so the workbook can get updated). that way we improve!

File: ios/platform-features/auto-layout/programmatical-constraints.workbook (hunk 3 above)
Comment (VincentDondain) [inline ios/platform-features/auto-layout/programmatical-constraints.workbook]: This is a bug in workbook, it runs all the code in the workbook, always. See: https://trello.com/c/tfzeDla6/913-cell-play-button-label-is-misleading-button-possibly-behaving-incorrectly.

Technically it should work so I don't really want to change my sentence here since it would go against the idea of workbooks (let's not have people try it in a classic app project, workbooks is the place to try and experiment).

It's on the workbooks team (:

**Comment by VincentDondain** [comment]
API doc issue so we can move away from referencing Apple doc and reference ours: https://github.com/xamarin/documentation/issues/2214

**Comment by VincentDondain** [comment]
@asb3993 can you approve so I can merge?
